src/core/ucc_lib.c

```diff
@@ -19,13 +20,219 @@ static ucc_config_field_t ucc_lib_config_table[] = {
     {NULL}
 };
 
+UCC_CONFIG_REGISTER_TABLE(ucc_lib_config_table, "UCC", NULL, ucc_lib_config_t)
+
+static inline ucc_status_t ucc_cl_component_is_loaded(const char *cl_name)
+{
+    if (NULL == ucc_get_component(&ucc_global_config.cl_framework, cl_name)) {
+        return UCC_ERR_NOT_FOUND;
+    } else {
+        return UCC_OK;
+    }
+}
+
+static inline ucc_status_t
+ucc_lib_config_cl_components_check(ucc_lib_info_t *lib,
+                                   ucc_config_names_array_t *cls)
+{
+    int i;
+    if (cls->count == 0) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+    if (cls->count == 1 && 0 == strcmp(cls->names[0], "all")) {
+        return UCC_OK;
+    }
+    for (i = 0; i < cls->count; i++) {
+        if (UCC_OK != ucc_cl_component_is_loaded(cls->names[i])) {
+            ucc_error("Required TL: \"%s\" (ucc_cl_%s.so) is not available\n",
+                      cls->names[i], cls->names[i]);
+            return UCC_ERR_NO_MESSAGE;
+        }
+    }
+    return UCC_OK;
+}
+
+static ucc_status_t ucc_lib_init_filtered(const ucc_lib_params_t *params,
+                                          const ucc_lib_config_t *config,
+                                          ucc_lib_info_t *lib)
+{
+    int                  n_cls = ucc_global_config.cl_framework.n_components;
+    uint64_t             supported_coll_types = 0;
+    ucc_cl_iface_t      *cl_iface;
+    ucc_cl_lib_t        *cl_lib;
+    ucc_cl_lib_config_t *cl_config;
+    ucc_status_t         status;
+    int                  i, specific_cls_requested;
+
+    lib->libs =
+        (ucc_cl_lib_t **)ucc_malloc(sizeof(ucc_cl_lib_t *) * n_cls, "cl_libs");
+    if (!lib->libs) {
+        ucc_error("failed to allocate %zd bytes for cl_libs",
+                  sizeof(ucc_cl_lib_t *) * n_cls);
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    ucc_assert(config->cls.count >= 1);
+    specific_cls_requested = (0 != strcmp(config->cls.names[0], "all"));
+    lib->n_libs_opened     = 0;
+    for (i = 0; i < n_cls; i++) {
+        cl_iface = ucc_derived_of(ucc_global_config.cl_framework.components[i],
+                                  ucc_cl_iface_t);
+        /* User requested specific list of CLs and current cl_iface is not part
+           of the list: skip it. */
+        if (specific_cls_requested &&
+            -1 == ucc_config_names_search(config->cls, cl_iface->super.name)) {
+            continue;
+        }
+        if (params->mask & UCC_LIB_PARAM_FIELD_THREAD_MODE &&
+            ((params->thread_mode & cl_iface->params.thread_mode) !=
+             params->thread_mode)) {
+            /* Requested THREAD_MODE is not supported by the CL:
+               1. If cls == "all" - just skip this CL
+               2. If specific CLs are requested: report error */
+            if (specific_cls_requested) {
```

Comment:
Wondering, if we really want to error out? Should it be a warning? The user asks for MULTI_THREAD and you can return SINGLE/FUNNELED. The MPI and OpenSHMEM models can tolerate this as long as it knows it.

Reply:
Well, yes, lets discuss this. We can implement either way. We can even have another env var option that controls the default behavior. 
We have 2 cases here: 1. User does not specify CLS, 2. User asks some specific list of CLS.
And then user asks for a thread mode. So we have 4 cases in total. Lets discuss in a call and define the behavior for all cases.